### PR TITLE
Switches the tagged metadata list from a skip list to a hash table

### PR DESCRIPTION
**Pull request by derobins**
#1928 Should be merged before this PR

### CODE REVIEW

**Comment by derobins** [comment]
This is a non-trivial performance increase in metadata-heavy workflows

File: src/H5private.h [hunk 1]
```diff
@@ -139,6 +139,15 @@
 #define R_OK 04
 #endif
 
+/* uthash is an external, header-only hash table implementation.
```
Comment (qkoziol) [inline src/H5private.h:142]: If this is brought in up here, the similar code in H5Ipkg.h and H5Dmpio.c should be removed.

Comment (qkoziol) [reply src/H5private.h:142]: Also - I understand that we don't want to redefine "HASH*" in the uthash.h header, but some sense of self-defense here if "HASH*" was already defined in a header that H5private has already included would be good.   Something like "HDcompile_assert(!defined(HASH_FIND));" before including the header?

**Comment by qkoziol** [comment]
Also suggest updating the version of uthash from the most recent one on github:   https://github.com/troydhanson/uthash   (Note different URL than in the header file in 'src')

**Comment by qkoziol** [comment]
Another thing - probably should define HASH_NONFATAL_OOM, so that the uthash code doesn't call exit() within the library.

**Comment by qkoziol** [comment]
And may want to see if uthash_fatal() could be defined to return an error from within the library.

**Comment by qkoziol** [comment]
Oops - accidentally clicked 'Close', sorry!

**Comment by derobins** [comment]
I'll probably sweep up any uthash changes in a separate PR that we can merge before this one. I like to keep to one "coding idea" per PR.

**Comment by derobins** [comment]
> Also suggest updating the version of uthash from the most recent one on github: https://github.com/troydhanson/uthash (Note different URL than in the header file in 'src')

Done (in #1928)

**Comment by derobins** [comment]
> Another thing - probably should define HASH_NONFATAL_OOM, so that the uthash code doesn't call exit() within the library.

Done (in #1928)

File: src/H5private.h (hunk 1 above)
Comment (derobins) [reply src/H5private.h:142]: Done (in #1928)

H5Ipkg.h was removed in this PR. I missed H5Dmpio.c.

**Comment by derobins** [comment]
> And may want to see if uthash_fatal() could be defined to return an error from within the library.

I added this to my to-do list, but I'm probably not going to address it here.

**Review by qkoziol — APPROVED** [review]

**Review by jhendersonHDF — APPROVED** [review]

**Pull request merged**